source/common/quic/platform/quiche_flags_impl.h

```diff
@@ -6,128 +6,17 @@
-  std::string help_;
-};
-
-// Concrete class for QUICHE protocol and feature flags, templated by flag type.
-template <typename T> class TypedFlag : public Flag {
-public:
-  TypedFlag(const char* name, T default_value, const char* help)
-      : Flag(name, help), value_(default_value), default_value_(default_value) {}
-
-  bool setValueFromString(const std::string& value_str) override;
-
-  void resetValue() override {
-    absl::MutexLock lock(&mutex_);
-    value_ = default_value_;
-  }
-
-  // Set flag value.
-  void setValue(T value) {
-    absl::MutexLock lock(&mutex_);
-    value_ = value;
-  }
-
-  // Return flag value.
-  T value() const {
-    absl::MutexLock lock(&mutex_);
-    return value_;
-  }
-
-private:
-  mutable absl::Mutex mutex_;
-  T value_ ABSL_GUARDED_BY(mutex_);
-  T default_value_;
-};
-
-// SetValueFromString specializations
-template <> bool TypedFlag<bool>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<int32_t>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<int64_t>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<double>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<std::string>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<unsigned long>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<unsigned long long>::setValueFromString(const std::string& value_str);
-
-// Flag declarations
-#define QUIC_FLAG(flag, ...) extern TypedFlag<bool>* flag;
-#include "quiche/quic/core/quic_flags_list.h"
-QUIC_FLAG(FLAGS_quic_reloadable_flag_spdy_testonly_default_false, false)
-QUIC_FLAG(FLAGS_quic_reloadable_flag_spdy_testonly_default_true, true)
-QUIC_FLAG(FLAGS_quic_restart_flag_spdy_testonly_default_false, false)
-QUIC_FLAG(FLAGS_quic_restart_flag_spdy_testonly_default_true, true)
-QUIC_FLAG(FLAGS_quic_reloadable_flag_http2_testonly_default_false, false)
-QUIC_FLAG(FLAGS_quic_reloadable_flag_http2_testonly_default_true, true)
-QUIC_FLAG(FLAGS_quic_restart_flag_http2_testonly_default_false, false)
-QUIC_FLAG(FLAGS_quic_restart_flag_http2_testonly_default_true, true)
-#undef QUIC_FLAG
-
-#define QUIC_PROTOCOL_FLAG(type, flag, ...) extern TypedFlag<type>* FLAGS_##flag;
-#include "quiche/quic/core/quic_protocol_flags_list.h"
-#undef QUIC_PROTOCOL_FLAG
+#include "common/quic/platform/flags_impl.h"
+#include "common/runtime/runtime_features.h"
 
 // |flag| is the global flag variable, which is a pointer to TypedFlag<type>.
 #define GetQuicheFlagImpl(flag) (quiche::flag)->value()
 
 // |flag| is the global flag variable, which is a pointer to TypedFlag<type>.
 #define SetQuicheFlagImpl(flag, value) (quiche::flag)->setValue(value)
 
-#define GetQuicheReloadableFlagImpl(module, flag) quiche::FLAGS_quic_reloadable_flag_##flag->value()
+#define GetQuicheReloadableFlagImpl(module, flag)                                                  \
```

Comment:
What prevents access of reloadable flags via the GetQuicheFlagImpl macro?

Reply:
Sorry I'm not 100% clear on what you mean.
You mean we should have some guard against directly accessing ReloadableFlag without going through Envoy runtime?

Reply:
It seems to me that GetQuicheFlagImpl(FLAGS_quic_reloadable_flag_foo) and GetQuicheReloadableFlagImpl(foo) generate equivalent code today, but after your change their behavior becomes different.

Going beyond these differences, suppose that we're attempting to build in a monolithic repo that hosts both Envoy and other binaries that also depend on QUICHE.  An implementation of GetQuicheReloadableFlagImpl that pays attention to Envoy runtime doesn't seem appropriate for this hypothetical monolithic repo setup.  Let's talk offline about how to reconcile QUICHE platform implementations when we need multiple QUICHE users to co-exist.